#366 Allow searching invoices by number

hmnhf opened the pull request:
Fixes #366

* Adds `invoices#restful_show_by_number` controller action (similar to `#restful_show`) and its route. ([demo](https://github.com/killbill/killbill-admin-ui/assets/6888728/21cc3d83-f0c6-43bf-90f9-10e19f43bfa9))
* Modifies `home#invoice_search` action and  `advanced_search_modal` partial to allow searching invoices by number. ([demo](https://github.com/killbill/killbill-admin-ui/assets/6888728/524232c0-2963-426e-84c3-59214a1ea1d2))
    * Renames `home#unsupported_external_key_search` to `#unsupported_search_field` and makes its message dynamic based on `search_by` value.
    * Makes "Advanced Search" modal's "Search by" options dynamic based on the selected object type. ([demo](https://github.com/killbill/killbill-admin-ui/assets/6888728/b63ce65a-69c1-4c17-8598-b47b559ac9f2))
        * Clears the selected "Search by" option if the newly selected "Object type" doesn't support it. ([demo](https://github.com/killbill/killbill-admin-ui/assets/6888728/6d1a0467-3fce-480d-aa44-fef0bcdb578d))
    * Removes an extra `INVOICE` item from "Advanced Search" object types. ([before](https://github.com/killbill/killbill-admin-ui/assets/6888728/1580f635-80af-4c56-bc74-a92a388aadfd) vs [after](https://github.com/killbill/killbill-admin-ui/assets/6888728/ccfaf4d4-8e68-4edf-838f-cea55afbce9c))
* Adds tests.

pierre reviewed (approved):
Looks great! Thanks also for adding these video demos, super helpful 👏 

Could you take a look at the CI errors?

hmnhf commented:
@pierre My pleasure!

I've fixed the failing tests and Rubocop violations and verified them locally.

For `Style/ArgumentsForwarding` violation in [lib/kaui.rb](https://github.com/killbill/killbill-admin-ui/blob/367aade3abf7a4bf10ea1767094e1e7c59c52e26/lib/kaui.rb#L209-L213), I first changed the code to use `...` as recommend [here](https://docs.rubocop.org/rubocop/cops_style.html#styleargumentsforwarding), and tests passed. But then I noticed the comment above the method override saying `ruby-1.8 compatibility` and although KAUI's README mentions it requires Ruby 3.2.2+, I just reverted the change and disabled the Rubocop check instead just to be safe. (`...` was introduced in Ruby 2.7)

We should be good to remove the method override altogether since Ruby 1.8 EOL was in 2014.
At least with Ruby 3.2.2 tests pass with it being removed. I wanted to know your opinion before doing something like that. (I think it should be done in a separate PR to be able to revert it easily if needed.)

pierre commented:
> We should be good to remove the method override altogether since Ruby 1.8 EOL was in 2014.
> At least with Ruby 3.2.2 tests pass with it being removed. I wanted to know your opinion before doing something like that. (I think it should be done in a separate PR to be able to revert it easily if needed.)

Agreed, let's get rid of this method altogether in a separate PR. 👍
